core/templates/services/exploration-improvements-task-registry.service.ts

```diff
@@ -250,64 +262,66 @@ export class ExplorationImprovementsTaskRegistryService {
           ...{target_id: newStateName},
         })
       ])),
-      oldStateTasks.supportingStateStats);
+      new SupportingStateStats(
+        this.expStats.getStateStats(newStateName),
+        oldStateTasks.supportingStats.answerStats,
+        oldStateTasks.supportingStats.cstPlaythroughIssues,
+        oldStateTasks.supportingStats.eqPlaythroughIssues,
+        oldStateTasks.supportingStats.misPlaythroughIssues));
 
     for (const newTask of newStateTasks) {
-      this.tasksByType.get(newTask.taskType).push(newTask);
+      if (newTask.isOpen()) {
+        this.pushOpenTask(newTask);
+      }
     }
     for (const oldTask of oldStateTasks) {
+      if (oldTask.isOpen()) {
+        this.popOpenTask(oldTask);
+      }
       oldTask.markAsObsolete();
     }
 
     this.tasksByState.set(newStateName, newStateTasks);
     // Map uses the reserved keyword delete as a method name.
     // eslint-disable-next-line dot-notation
     this.tasksByState.delete(oldStateName);
-    this.expStats = this.expStats.createNewWithStateRenamed(
-      oldStateName, newStateName);
   }
 
-  onChangeInteraction(stateName: string): void {
-    this.tasksByState.get(stateName).refresh(this.expStats, this.config);
+  onStateInteractionSaved(state: State): void {
+    this.refreshStateTasks(state.name);
   }
 
   getOpenHighBounceRateTasks(): HbrTask[] {
-    return <HbrTask[]> this.tasksByType.get(
-      ImprovementsConstants.TASK_TYPE_HIGH_BOUNCE_RATE)
-      .filter(t => t.isOpen());
+    return <HbrTask[]> this.openTasksByType.get(
+      ImprovementsConstants.TASK_TYPE_HIGH_BOUNCE_RATE);
   }
 
   getOpenIneffectiveFeedbackLoopTasks(): IflTask[] {
-    return <IflTask[]> this.tasksByType.get(
-      ImprovementsConstants.TASK_TYPE_INEFFECTIVE_FEEDBACK_LOOP)
-      .filter(t => t.isOpen());
+    return <IflTask[]> this.openTasksByType.get(
+      ImprovementsConstants.TASK_TYPE_INEFFECTIVE_FEEDBACK_LOOP);
   }
 
   getOpenNeedsGuidingResponsesTasks(): NgrTask[] {
-    return <NgrTask[]> this.tasksByType.get(
-      ImprovementsConstants.TASK_TYPE_NEEDS_GUIDING_RESPONSES)
-      .filter(t => t.isOpen());
+    return <NgrTask[]> this.openTasksByType.get(
+      ImprovementsConstants.TASK_TYPE_NEEDS_GUIDING_RESPONSES);
   }
 
   getOpenSuccessiveIncorrectAnswersTasks(): SiaTask[] {
-    return <SiaTask[]> this.tasksByType.get(
-      ImprovementsConstants.TASK_TYPE_SUCCESSIVE_INCORRECT_ANSWERS)
-      .filter(t => t.isOpen());
+    return <SiaTask[]> this.openTasksByType.get(
+      ImprovementsConstants.TASK_TYPE_SUCCESSIVE_INCORRECT_ANSWERS);
   }
 
-  getSupportingStateStats(task: ExplorationTask): SupportingStateStats {
-    if (!this.tasksByState.has(task.targetId)) {
-      throw new Error('Unregistered task has no supporting stats');
-    }
-    return this.tasksByState.get(task.targetId).supportingStateStats;
+  makeStateTasksDictionary(): Dictionary<StateTasks> {
+    return fromPairs([...this.tasksByState]);
```

Comment:
Why are we introducing a new "Dictionary" type, when we already have "Object" or "Map"?

Also, I'm largely confused about what this is doing. Could you explain? Conceptually I don't understand why you are converting key-value pairs to a list and then back again. 

Finally I believe we only use lodash for cloneDeep. Can we try not to use it for other things, if it's not necessary to do so? It just makes things harder for new devs to understand.

Reply:
> Why are we introducing a new "Dictionary" type, when we already have "Object" or "Map"?

So `Dictionary<T>` is just a type alias for `{[index: string]: T}`. I used it because the latter is less readable IMO.

> Also, I'm largely confused about what this is doing. Could you explain? Conceptually I don't understand why you are converting key-value pairs to a list and then back again.

AngularJS can only iterate over objects and arrays, it doesn't work with `Map`. To transform the map into an object, I used the `fromPairs` function which does just that.

> Finally I believe we only use lodash for cloneDeep. Can we try not to use it for other things, if it's not necessary to do so? It just makes things harder for new devs to understand.

Yeah... I *wanted* to use [Object.fromEntries](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/fromEntries), but it doesn't look like we're using a new-enough ECMA version, so I settled for lodash's `fromPairs` instead. This task seems mundane/common enough to deserve a third-party library call, but not sure which to use.

---

I went ahead and got rid of the alias/fromPairs call, since it's not too big of a deal to me.